format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/PathTrackingDelimitedInputFormat.java

```diff
@@ -69,66 +69,72 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
         if (delegate.nextKeyValue()) {
           // skip to next if the current record is header
           if (skipHeader && delegate.getCurrentKey().get() == 0L) {
-            return delegate.nextKeyValue();
+            if (!delegate.nextKeyValue()) {
+              return false;
+            }
+          }
+          // this logic must be in nextKeyValue to prevent multiple calls to getCurrentValue
+          // from advancing the delegate reader
+          String delimitedString = delegate.getCurrentValue().toString();
+          builder = StructuredRecord.builder(schema);
+          Iterator<Schema.Field> fields = schema.getFields().iterator();
+          Iterator<String> splitsIterator = getSplitsIterator(enableQuotesValue, delimitedString, delimiter);
+          int dataFieldsCount = 0;
+          while (splitsIterator.hasNext()) {
+            dataFieldsCount++;
+            String part = splitsIterator.next();
+            if (!fields.hasNext()) {
+              while (splitsIterator.hasNext()) {
```

Comment:
thinking about this a bit more, there's a bad data scenario where there is an unenclosed quote. In that case, this will loop through the entire file, which can take a long time.

It would be a little better to pass this iterator to the handlerImproperString() method. In there, we can add some limit on the number of iterations (ex: 100), after which we throw an exception with a more specific message about there likely being an unenclosed quote.

Reply:
this can also be done in a follow up PR if we just want to get the core logic merged first

Reply:
sure, that can be done later.